src/porepy/compositional/compositional_mixins.py

```diff
@@ -1222,106 +1217,102 @@ def dependencies_of_phase_properties(
+
+                        dependencies = [self.pressure]
+                        if self._has_unified_equilibrium:
+
+                            dependencies +=  [self.temperature] + [
+                                phase.extended_fraction_of[component]
+                                for component in phase
+                                if self.has_independent_extended_fraction(
+                                    component, phase
+                                )
+                            ]
+
+                        elif self._has_equilibrium:
+
+                            dependencies += [self.temperature] + [
+                                phase.partial_fraction_of[component]
+                                for component in phase
+                                if self.has_independent_partial_fraction(
+                                    component, phase
+                                )
+                            ]
+
+                        else:
 
-    def density_of_phase(self, phase: Phase) -> pp.ad.SurrogateFactory:
+                            dependencies += [
+                                component.fraction
+                                for component in self.fluid.components
+                                if self.has_independent_fraction(component)
+                            ]
+
+                        return dependencies
+
+        """
+        return []
+
+    def density_of_phase(self, phase: Phase) -> ExtendedDomainFunctionType:
         """This base method returns the density of a ``phase`` as a
-        :class:`~porepy.numerics.ad.surrogate_operator.SurrogateFactory`.
+        :class:`~porepy.numerics.ad.surrogate_operator.SurrogateFactory`,
+        if :meth:`dependencies_of_phase_properties` has a non-empty return value.
 
-        It is populated by the solution strategy, depending on how
-        the equilibrium conditions and constitutive laws are implemented.
+        Otherwise it returns an empty function, raising an error when called
+        (missing mixin of constitutive laws).
 
         The phase density (like all thermodynamic properties) is a dependent quantity.
 
         """
-        return pp.ad.SurrogateFactory(
-            f"phase_{phase.name}_density",
-            self.mdg,
-            self.dependencies_of_phase_properties(phase),
-        )
+        name = f"phase_{phase.name}_density"
+        dependencies = self.dependencies_of_phase_properties(phase)
+        if dependencies:
+            return _get_surrogate_factory_as_property(name, self.mdg, dependencies)
+        else:
+            return _no_property_function
 
-    def specific_volume_of_phase(
-        self, phase: Phase
-    ) -> Callable[[pp.SubdomainsOrBoundaries], pp.ad.Operator]:
+    def specific_volume_of_phase(self, phase: Phase) -> ExtendedDomainFunctionType:
         """The specific volume of the phase is returned as a function calling the
```

Comment:
Please add parameters and return here and below.

Reply:
Parameters and return values added.